### ISSUE TITLE
Source handler questions and issues

### ISSUE THREAD

**Issue by dmlap** [videojs/video.js#2323]
I'm in the process of converting videojs-contrib-hls from a tech to a source handler. contrib-hls is a pretty complicated extension due to its interaction with the Flash tech so it may not be representative of the sort of problems other source handler authors would encounter. Nonetheless, I thought it would be helpful to list all the questions I hit along the way for discussion purposes. I'm going to fill in this issue with notes as I work through the migration.

**Comment by dmlap** [comment]
- Source handlers aren't tied to the lifecycle of the tech they're registered to. Any interesting source handler is going to need to manage some state. The source handler object shares the page lifetime since it's registered on the tech class.
- Source handlers are responsible for tracking their tech themselves. Doable, just seems like something all source handlers will need a persistent reference to

**Comment by dmlap** [comment]
- How do you pass initialization options to a source handler?

**Comment by dmlap** [comment]
- We don't support `vjs-has-started` at the tech level. Is it possible to find out whether a video has been played at all or not? (`video.played`?)

**Comment by dmlap** [comment]
- techs do not currently have an `error` property so it's impossible to trigger a custom error

**Comment by dmlap** [comment]
- tech level events are explicitly re-dispatched  *and* "propagated" to the player, `durationchange` for instance, causing duplicate event triggers.

**Comment by dmlap** [comment]
- It is not possible to implement a native-looking `seekable` for [live streams with Media Source Extensions](https://w3c.github.io/media-source/#htmlmediaelement-extensions) and there is no way for a source handler to override `seekable` behavior at the tech level.

**Comment by heff** [comment]
> Source handlers are responsible for tracking their tech themselves. Doable, just seems like something all source handlers will need a persistent reference to

The simplest source handler is the native source handler, which doesn't need to store a reference to the tech. If there's common boilerplate for adaptive source handlers, I think we could create an AdaptiveSourceHandler class or something similar that sourcehandlers could subclass, to share the boilerplate, instead of making the source handler interface itself more complicated.

> How do you pass initialization options to a source handler?

That's a good question. Can you give an example?

> We don't support vjs-has-started at the tech level. Is it possible to find out whether a video has been played at all or not? (video.played?)

Yeah, `played` would be the appropriate property for determining that state. It should work on the HTML5 side but we may need to build something for other techs.

> techs do not currently have an error property so it's impossible to trigger a custom error

A custom error that kills the player process? What's the example you're working with? Could endOfStream be related?

> tech level events are explicitly re-dispatched and "propagated" to the player, durationchange for instance, causing duplicate event triggers.

I'm pretty sure that's a bug. With the Tech 2.0 changes we flipped the model to have [the player listen for events on the tech](https://github.com/videojs/video.js/blob/004b1e8f7bfbae1749e7b141b9ccd45b3d1a5bca/src/js/player.js#L486). But if you're seeing that then our event system may also be automatically bubbling events, even though media element events should have bubble=false.

> It is not possible to implement a native-looking seekable for live streams with Media Source Extensions and there is no way for a source handler to override seekable behavior at the tech level.

Could use some more details here. What does the native seekable return in this case? Could it be a spec bug?

**Comment by gkatsev** [comment]
One way of passing the options to the source handler is to just set it as an option on the appropriate tech and then have the source handler read it from the tech's options.

**Comment by dmlap** [comment]
@heff: 
>> techs do not currently have an error property so it's impossible to trigger a custom error

> A custom error that kills the player process? What's the example you're working with? Could endOfStream be related?

I need to trigger a MEDIA_ERR_NETWORK if one of the m3u8s are unavailable

>> It is not possible to implement a native-looking seekable for live streams with Media Source Extensions and there is no way for a source handler to override seekable behavior at the tech level.

> Could use some more details here. What does the native seekable return in this case? Could it be a spec bug?

The spec says that `seekable` should return the same thing as `buffered` when duration is Infinity. Safari on iOS and desktop makes seekable any of the segments listed in the m3u8 (minus a couple at the end so you don't bang up against the live point), which is significantly more than `buffered`.

**Comment by heff** [comment]
> I need to trigger a MEDIA_ERR_NETWORK if one of the m3u8s are unavailable

Ok yeah, I don't have an immediate solution but we'll need one.

> The spec says that seekable should return the same thing as buffered when duration is Infinity. Safari on iOS and desktop makes seekable any of the segments listed in the m3u8 (minus a couple at the end so you don't bang up against the live point), which is significantly more than buffered.

Can we not make the spec version work, or are you saying you want to copy the iOS version?

**Comment by dmlap** [comment]
> Can we not make the spec version work, or are you saying you want to copy the iOS version?

The HTML spec defines a live seekable which is more in line with the iOS experience. I don't think it would be possible to build something like Live DVR controls on top of MSE's definition without outside-the-spec info on the manifest. It sounds like it's worth raising against the MSE spec. I definitely think we should be shooting for the iOS flavor downstream from video.js, though.

**Comment by heff** [comment]
> It sounds like it's worth raising against the MSE spec

Yeah, does to me too. Want to [submit an issue](https://github.com/w3c/media-source)? I think they'd appreciate the use case info.

I guess in the mean time we'll have to update the tech seekable to check if the source handler wants to provide that info...

In `Tech.withSourceHandlers`...
```
let origSeekable = _Tech.prototype.seekable;
_Tech.prototype.seekable = function(){
  if (this.sourceHandler_ && this.sourceHandler_.seekable) {
    return this.sourceHandler_.seekable();
  } else {
    return origSeekable && origSeekable.call(this);
  }
};
```
I don't love it but that's the best I can come up with.

**Comment by dmlap** [comment]
We could add a setSeekable() a la duration but it changes frequently during playback and that would probably be a pain for source handlers to keep synced. I filed the [MSE issue](https://github.com/w3c/media-source/issues/5).

**Comment by dmlap** [comment]
- `progress` triggered on the tech is bubbling (just like `durationchange`)

**Comment by dmlap** [comment]
- `addTextTrack()` only exists on the player so source handlers can't add in-band metadata tracks

**Comment by dmlap** [comment]
- custom tech level events do not get re-dispatched at the player level by default

**Comment by heff** [comment]
> custom tech level events do not get re-dispatched at the player level by default

What events did you have in mind? Ideally we'd find an event that could be standard across techs and have the player start listening for that.

**Comment by dmlap** [comment]
We have some HLS-specific events in contrib-hls that previously were fired on the player for easy access. There's a `mediachange` event, for instance, that indicates the quality level being buffered has changed.

**Comment by heff** [comment]
`mediachange` seems a little too ambiguous at the player level, but I could see creating a `renditionchange` or other named event that the player just proxies from the tech.

Otherwise, can the event be fired on the tech now instead?

**Comment by dmlap** [comment]
They could be but it would be a backwards-incompatible change.

**Comment by dmlap** [comment]
- Seek positions should be clamped to the current seekable range for the video. The tech will have to enforce this since Media Source Extensions narrows `seekable` to the buffered region for live video.

**Comment by heff** [comment]
@dmlap where is the HLS work now in relation to 5.0? Blocked, good to go, or somewhere in between?

I ultimately need to know if we can release 5.0 with source handlers as they are, or what's the minimum we need to do to get there.

A lot of the discussion items appear to have closed pull requests associated with them, so it seems like we should start opening issues for the left over items instead of trying to track it all here. We could create source handlers milestone if that would help keep them organized.

Here's what I think is left:
- Define how source handlers should receive configuration options
- Define how a native-looking seekable would be implemented. I think this includes (your last comment)[https://github.com/videojs/video.js/issues/2323#issuecomment-132369641]
- Define a player-level event for tech-level rendition switching (to cover the HLS `mediachange` use case)

**Comment by heff** [comment]
Per @dmlap in chat
> if you pulled in raynos/xhr, we have some tiny changes to make, otherwise, gtg

I feel like the open points here could be done in minor releases, and we need to get 5.0 closed, so moving this to the 5.1 milestone.

**Comment by gkatsev** [comment]
I think these are mostly settled and we still have https://github.com/videojs/video.js/issues/2268
